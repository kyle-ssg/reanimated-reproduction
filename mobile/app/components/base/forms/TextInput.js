/**
 * Created by kylejohnson on 14/11/2015.
 */

import InputMask from 'inputmask-core';

import React, {Component, PropTypes} from 'react';

const TextInput = class extends Component {
    displayName: 'TextInput'

    componentWillMount() {
        this.animation = new Animated.Value(0)
    }

    clear = () => {
        this.refs.input.clear();
    };

    blur = () => {
        this.refs.input.blur();
    };

    focus = () => {
        this.refs.input.focus();
    };

    onFocus = () => {
        Animated.timing(this.animation, {
            toValue: 1,
            duration: 300,
            useNativeDriver: true, // <-- Add this
            easing: Easing.cubic
        }).start();
        this.props.onFocus && this.props.onFocus();
    };

    onBlur = () => {
        Animated.timing(this.animation, {
            toValue: 0,
            duration: 300,
            useNativeDriver: true, // <-- Add this
            easing: Easing.cubic
        }).start();
        this.props.onBlur && this.props.onBlur();
    };

    onChangeText = (text) => {
        if (this.props.mask) {
            // Masking
            if (!this.mask) {
                // Create new mask
                this.mask = new InputMask({
                    pattern: this.props.mask,
                    formatCharacters: {
                        'a': {
                            validate(char) {
                                return /[ap]/.test(char);
                            }
                        },
                        'm': {
                            validate(char) {
                                return /\w/.test(char);
                            },
                            transform() {
                                return 'm';
                            }
                        }
                    }
                });
            }

            if (text.length > this.mask.selection.start) {
                // Character(s) were typed, ignore if text exceeds length of mask
                if (this.mask.selection.start === this.props.mask.length) {
                    return;
                }

                // It does not, extract the character(s) that were added
                text = text.slice(this.mask.selection.start);

                // Add it to the input mask
                if (text.length > 1) {
                    this.mask.paste(text);
                }
                else {
                    // Perform additional inputs to skip non-pattern characters. Input will be converted
                    // to the non-pattern character.
                    while (!this.isMaskPatternChar(this.props.mask[this.mask.selection.start]) &&
                    this.mask.selection.start !== this.props.mask.length) {
                        // On failure abort loop as cursor position will not change
                        if (!this.mask.input(text)) {
                            break;
                        }
                    }

                    this.mask.input(text);
                }
            }
            else if (text.length < this.mask.selection.start) {
                // Character(s) were deleted, delete up to current length
                while (this.mask.selection.start != text.length)
                    this.mask.backspace();

                // Check whether more backspaces are required until we reach a pattern char or nothing is left
                while (this.mask.selection.start && !this.isMaskPatternChar(this.props.mask[this.mask.selection.start - 1])) {
                    this.mask.backspace();
                }
            }

            const value = this.mask.getValue().slice(0, this.mask.selection.start);
            console.log(value)
            // Update text
            this.props.onChangeText && this.props.onChangeText();
        }
        else {
            // No masking, just update text
            this.props.onChangeText && this.props.onChangeText(text);
        }
    }

    isMaskPatternChar(char) {
        if (!char || char.length !== 1) {
            return false;
        }

        return char === '1' || char === 'a' || char === 'A' || char === '*' || char === '#';
    }


    render() {
        return (
            <View>
                <ReactNative.TextInput
                    {...this.props}
                    onFocus={this.onFocus}
                    onBlur={this.onBlur}
                    onChangeText={this.onChangeText}
                    style={[this.props.style, Styles.textInput]}
                    value={this.props.value}
                />
                <Animated.View style={[{
                    marginTop: -styleVariables.inputBorderWidth,
                    transform: [{scaleX: this.animation}],
                    backgroundColor: colour.activeBorder, height: 1 / PixelRatio.get() * 4
                }]}/>
            </View>
        );
    }
};

TextInput.propTypes = {};

TextInput.propTypes = {
    value: OptionalString,
    placeholder: OptionalString,
    editable: OptionalBool,
    multiline: OptionalBool,
    maxLines: OptionalNumber,
    minLines: OptionalNumber,
    mask: OptionalString,
    onChangeText: OptionalFunc,
    height: OptionalNumber,
    style: React.PropTypes.any,
    secureTextInput: OptionalBool,
    disabled: OptionalBool,
    keyboardType: OptionalBool,
    onSubmit: OptionalFunc,
    onFocus: OptionalFunc,
    textStyle: oneOfType([OptionalObject, OptionalNumber]),
};
module.exports = TextInput;