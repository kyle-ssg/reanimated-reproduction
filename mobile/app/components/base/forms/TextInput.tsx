/**
 * Created by kylejohnson on 14/11/2015.
 */

import ReactNative from 'react-native'

import React, { FunctionComponent, useRef, useState } from 'react' // we need this to make JSX compile

export type TextInputProps = ReactNative.TextInputProps & {
  title?: string
  style?: ReactNative.StyleProp<ReactNative.ViewStyle>
  invalid?: boolean
  disabled?: boolean
  textStyle?: ReactNative.TextStyle
  isLight?: boolean
  icon?: string
  iconColour?: string
}

const TextInput: FunctionComponent<TextInputProps> = (props) => {
  const inputRef = useRef<ReactNative.TextInput>()
  const [isSecureTextEntry, setIsSecureTextEntry] = useState<boolean>(
    props.secureTextEntry,
  )

  const clear = () => {
    inputRef.current.clear()
  }

  const blur = () => {
    inputRef.current.blur()
  }

  const focus = () => {
    inputRef.current.focus()
  }

  const onFocus = (e) => {
    if (props.onFocus) {
      props.onFocus(e)
    }
  }

  const onBlur = (e) => {
    if (props.onBlur) {
      props.onBlur(e)
    }
  }

  const onChangeText = (text) => {
    if (!props.onChangeText) {
      return
    }
    props.onChangeText(text)
  }

  return (
    <>
      {props.title && (
        <FormGroup>
          {props.isLight ? (
            <>
              <Text style={[Styles.textInputTitleLight, props.textStyle]}>
                {props.title}
              </Text>
            </>
          ) : (
            <Text style={[Styles.textInputTitleDark, props.textStyle]}>
              {props.title}
            </Text>
          )}
        </FormGroup>
      )}
      <View>
        <ReactNative.TextInput
          {...props}
          secureTextEntry={isSecureTextEntry}
          onFocus={onFocus}
          editable={!props.disabled && props.editable !== false}
          onBlur={onBlur}
          onChangeText={onChangeText}
          style={[
            Styles.textInput,
            props.style,
            props.invalid && Styles.textInputError,
          ]}
          value={props.value}
          testID={props.testID}
          ref={inputRef}
          blurOnSubmit={Platform.OS === 'ios' && !props.multiline}
          placeholderTextColor={
            props.invalid ? palette.danger : 'rgba(11,16,46,0.5)'
          }
          selectionColor={palette.primary}
          icon={props.icon}
          iconColour={props.icon}
          multiline={props.multiline}
          textAlignVertical={props.textAlignVertical}
        />
        {props.icon ? (
          <>
            <FA5Pro
              solid
              style={Styles.textInputIcon}
              name={props.icon}
              size={20}
              color={props.iconColour || palette.primary}
            />
          </>
        ) : null}

        {props.secureTextEntry ? (
          <TouchableOpacity
            onPress={() => setIsSecureTextEntry(!isSecureTextEntry)}
            style={[{ position: 'absolute', width: 50, right: 0, height: 50 }]}
          >
            <FA5Pro
              style={Styles.textInputIcon}
              name={isSecureTextEntry ? 'eye' : 'eye-slash'}
              size={20}
              color={palette.primary}
            />
          </TouchableOpacity>
        ) : null}
      </View>
    </>
  )
}

export default TextInput
