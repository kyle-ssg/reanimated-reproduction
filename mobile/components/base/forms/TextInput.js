/**
 * Created by kylejohnson on 14/11/2015.
 */
module.exports = Component({
    clear: function() {
        this.refs.input.clear();
    },
    blur: function() {
        this.refs.input.blur();
    },
    focus: function() {
        this.refs.input.focus();
    },
    getInitialState: function() {
        return {
            height: new Animated.Value(this.getHeight(this.props.text || ""))
        }
    },
    componentWillReceiveProps: function(newProps) {
        if (this.props.multiline && this.props.maxLines && newProps.value != this.props.value) {
            Animated.timing(this.state.height, {
                toValue: this.getHeight(newProps.value),
                duration: 200
            }).start();
        }
    },
    onChangeText: function(text) { //use more familiar onChange
        if (this.props.onChangeText) {
            this.props.onChangeText(text);
        } else if (this.props.onChange) {
            this.props.onChange(text);
        }
    },
    getHeight: function(text) {
        var lines = text.match(/\n/g);
        var linesToShow = Math.min((Math.max(lines && lines.length || 0), (this.props.minLines || 0)), this.props.maxLines);
        if (isNaN(linesToShow)) {
            linesToShow = 0;
        }
        return (linesToShow * 16) +  (this.props.height || styleVariables.inputHeight) ;
    },
    render: function() {
        return  <Animated.View style={[Styles.inputContainer, this.props.style, {height:this.state.height}]}>
            <ReactNative.TextInput
                placeholderTextColor={styleVariables.placeholderTextColor}
                ref={"input"}
                onChangeText={this.onChangeText}
                multiline={this.props.multiline}
                secureTextEntry={this.props.secureTextInput}
                onFocus={this.props.onFocus}
                value={this.props.value}
                placeholder={this.props.placeholder}
                editable={this.props.editable || !this.props.disabled}
                style={[Styles.input, this.props.textStyle]}
                onSubmitEditing={this.props.onSubmitEditing}
                keyboardType={this.props.keyboardType}/>
        </Animated.View>
    }
})