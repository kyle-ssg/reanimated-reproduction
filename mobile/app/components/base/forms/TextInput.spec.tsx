import * as React from 'react'
import TextInput from './TextInput'
import renderer from 'react-test-renderer'

describe('<TextInput />', () => {
  it(`renders correctly`, () => {
    const tree = renderer.create(<TextInput />)
    expect(tree).toMatchSnapshot()
  })
})
