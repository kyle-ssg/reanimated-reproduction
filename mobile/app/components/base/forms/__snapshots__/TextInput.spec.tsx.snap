// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> renders correctly 1`] = `
<View>
  <TextInput
    allowFontScaling={true}
    blurOnSubmit={true}
    editable={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholderTextColor="rgba(11,16,46,0.5)"
    rejectResponderTermination={true}
    selectionColor="#1AC0C6"
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "fontFamily": "System",
          "fontSize": 20,
          "height": 48,
          "paddingLeft": 10,
        },
        undefined,
        undefined,
      ]
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
